pkg/utils/managedresources/registry.go

```diff
@@ -0,0 +1,153 @@
+// types that will later be added to the registry.
+func NewRegistry(scheme *runtime.Scheme, codec serializer.CodecFactory, serializer *json.Serializer) *Registry {
+	var groupVersions []schema.GroupVersion
+	for k := range scheme.AllKnownTypes() {
+		groupVersions = append(groupVersions, k.GroupVersion())
+	}
+
+	return &Registry{
+		codec:        codec.CodecForVersions(serializer, serializer, schema.GroupVersions(groupVersions), schema.GroupVersions(groupVersions)),
+		nameToObject: make(map[string]*object),
+	}
+}
+
+// Add adds the given object the registry. It computes a filename based on its type, namespace, and name. It serializes
+// the object to YAML and stores both representations (object and serialization) in the registry.
+func (r *Registry) Add(obj runtime.Object) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if obj == nil || reflect.ValueOf(obj) == reflect.Zero(reflect.TypeOf(obj)) {
+		return nil
+	}
+
+	objectName, err := r.objectName(obj)
+	if err != nil {
+		return err
+	}
+	filename := objectName + ".yaml"
+
+	if _, ok := r.nameToObject[filename]; ok {
+		return fmt.Errorf("duplicate filename in registry: %q", filename)
+	}
+
+	serializationYAML, err := runtime.Encode(r.codec, obj)
+	if err != nil {
+		return err
+	}
+
+	r.nameToObject[filename] = &object{
+		obj:           obj,
+		serialization: serializationYAML,
+	}
+
+	return nil
+}
+
+// SerializedObjects returns a map whose keys are filenames and whose values are serialized objects.
+func (r *Registry) SerializedObjects() map[string][]byte {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	out := make(map[string][]byte, len(r.nameToObject))
+	for name, object := range r.nameToObject {
+		out[name] = object.serialization
+	}
+	return out
+}
+
+// AddAllAndSerialize calls Add() for all the given objects before calling SerializedObjects().
+func (r *Registry) AddAllAndSerialize(objects ...runtime.Object) (map[string][]byte, error) {
+	for _, resource := range objects {
+		if err := r.Add(resource); err != nil {
+			return nil, err
+		}
+	}
+	return r.SerializedObjects(), nil
+}
+
+// RegisteredObjects returns a map whose keys are filenames and whose values are objects.
+func (r *Registry) RegisteredObjects() map[string]runtime.Object {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	out := make(map[string]runtime.Object, len(r.nameToObject))
+	for name, object := range r.nameToObject {
+		out[name] = object.obj
+	}
+	return out
+}
+
+// String returns the string representation of the registry.
+func (r *Registry) String() string {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	out := make([]string, 0, len(r.nameToObject))
+	for name, object := range r.nameToObject {
+		out = append(out, fmt.Sprintf("* %s:\n%s", name, object.serialization))
+	}
+	return strings.Join(out, "\n\n")
+}
+
+func (r *Registry) objectName(obj runtime.Object) (string, error) {
+	var (
+		typeOf   = strings.Split(reflect.TypeOf(obj).String(), ".")
+		typeName = strings.ToLower(typeOf[len(typeOf)-1])
```

Comment:
Why did you choose to use reflection directly here, instead of reusing the scheme for checking the GVK of the object?
You can also use `apiutil.GVKForObject`, if convenience is a concern...

Reply:
No specific reason, I simply wasn't thinking of this api utility function, thanks for mentioning it. I'll improve it.